File: AutonomousSport/src/models/Room.js
const TAG = 'Room';
export default class Room {
  constructor(roomJson) {
    console.log(TAG, ' constructor roomJson = ', roomJson);
    this.id = roomJson?.id || -1;
    this.userId = roomJson?.userId || 0; // is mentor
    this.createdAt = roomJson?.createdAt || '';
    this.updatedAt = roomJson?.updatedAt || '';
    this.deletedAt = roomJson?.deletedAt || null;
    this.name = roomJson?.name || '';
    this.photo = roomJson?.photo || '';
    this.cover = roomJson?.cover || '';
    this.session = roomJson?.session || '';
    this.token = roomJson?.token || '';
    this.status = roomJson?.status || 0;
    this.loop = roomJson?.loop || 0;
    this.RoomPlayers = roomJson?.RoomPlayers || [];
    this.Map = roomJson?.Map || {};
    this.mapId = roomJson?.mapId || 0;
    this.miles = roomJson?.miles || 0;
  }

  getPathOfMap = (): [] => {
    try {
      const json = JSON.parse(this.Map?.datapoints);
      return json['points'] || [];
    } catch (error) {}
    return [];
  };

  getMapSize = (): { width: 0, height: 0 } => {
    const map = this.Map || { width: 0, height: 0 };
    return {
      width: map.width,
      height: map.height
    };
  };
  toJSON() {
    return {
      id: this.id,
      userId: this.userId,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
      deletedAt: this.deletedAt,
      name: this.name,
      session: this.session,
      token: this.token,
      photo: this.photo,
      cover: this.cover,
      Map: this.Map,
      RoomPlayers: this.RoomPlayers || [],
      status: this.status,
      loop: this.loop,
      mapId: this.mapId,
      miles: this.miles
    };
  }
}

File: Backend/models/room.go
package models

import (
	"time" 
	// "net/url"
	"errors"
	// "sort"
	// "io"
	//"fmt" 
	"../config"
	"github.com/jinzhu/gorm"  
)
 
// Room ...
type Room struct { 
	ID         int        `json:"id" gorm:"primary_key"`
	UserId    int    `json:"userId"`
	CreatedAt *time.Time `json:"createdAt, omitempty"`
	UpdatedAt *time.Time `json:"updatedAt, omitempty"`
	DeletedAt *time.Time `json:"deletedAt, omitempty" sql:"index"`  
	Name string `json:"name, omitempty" gorm:"not null; type:varchar(100)"` 
	Photo string `json:"photo, omitempty" gorm:"not null; type:varchar(500)"` 
	Session     string `json:"session, omitempty" gorm:"type:varchar(100);unique;unique_index"`
	Token string `json:"token, omitempty" gorm:"not null; type:varchar(500)"`  
	Status     int        `json:"status, omitempty" gorm:"not null"`  
	MapId	   int        `json:"mapId, omitempty" gorm:"not null"`  
	Loop	   int        `json:"loop, omitempty" gorm:"not null"`  
	Miles	   float64        `json:"miles, omitempty" gorm:"not null"`  

}

// TableName set Room's table name to be `profiles`
func (Room) TableName() string {
	return "rooms"
}

    

// FetchAll ...
func (u *Room) FetchAll() []Room {
	db := config.GetDatabaseConnection()

	var rooms []Room
	db.Find(&rooms)
	//db.Limit(3).Find(&users)
	return rooms
}

// FetchById ...
func (u *Room) FetchById(db *gorm.DB) error {
 
	if err := db.Where("id = ?", u.ID).Find(&u).Error; err != nil {
		return errors.New("Could not find the Room")
	} 
	return nil
}
// FetchById ...
func (u *Room) FetchBySession(db *gorm.DB) error {
	  
	if err := db.Where("session = ?", u.Session).Find(&u).Error; err != nil {
		return errors.New("Could not find the Room")
	} 
	return nil
}
// Create ...
func (u *Room) Create(db *gorm.DB) error {
	//db := config.GetDatabaseConnection()
	// Validate record
	if !db.NewRecord(u) { // => returns `true` as primary key is blank
		return errors.New("New records can not have primary key id")
	}

	if err := db.Create(&u).Error; err != nil {
		return errors.New("Could not create Room")
	}
	
	//InitBalance...
	// db.Create(&Balance{UserId: u.ID, Name:"ETH", Balance:0})
	// db.Create(&Balance{UserId: u.ID, Name:"BTC", Balance:0})
	// db.Create(&Balance{UserId:  u.ID, Name:"LTC", Balance:0}) 

	return nil
}

// Save ...
func (u *Room) Save(db *gorm.DB) error {
 
	if db.NewRecord(u) {
		if err := db.Create(&u).Error; err != nil {
			return errors.New("Could not create Room")
		}
	} else {
		if err := db.Save(&u).Error; err != nil {
			return errors.New("Could not update Room")
		}
	}

	return nil
}

// Delete ...
func (u *Room) Delete() error {
	db := config.GetDatabaseConnection()

	if err := db.Delete(&u).Error; err != nil {
		return errors.New("Could not find the Room")
	}

	return nil
}
